### ISSUE TITLE
Switching between encrypt/decryption on EVP_CIPHER_CTX silently computes wrong results, rather than errors

### ISSUE THREAD

**Issue by alex** [openssl/openssl#22070]
An `EVP_CIPHER_CTX` can be used for either encryption or decryption. However, with _some_ ciphers, using a single context to do both can produce wrong results. Here's a simple example with AES-OCB:

```c
#include <stdbool.h>
#include <string.h>

#include <openssl/err.h>
#include <openssl/evp.h>

void openssl_assert(bool x, char *op) {
  if (!x) {
    printf("%s failed\n", op);
    ERR_print_errors_fp(stderr);
    exit(1);
  } else {
    printf("%s succeeded\n", op);
  }
}

int main() {
  int res, len;

  unsigned char key[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
  unsigned char nonce[] = {187, 170, 153, 136, 119, 102, 85, 68, 51, 34, 17, 4};
  unsigned char pt[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
  unsigned char ct[] = {87,  29,  83,  91,  96, 178, 119, 24,  139, 229, 20,
                        113, 112, 169, 162, 44, 58,  215, 164, 255, 56,  53,
                        184, 197, 112, 28,  28, 206, 200, 252, 51,  88};
  unsigned char aad[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};

  const EVP_CIPHER *cipher = EVP_aes_128_ocb();
  openssl_assert(cipher != NULL, "cipher =");

  EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
  openssl_assert(ctx != NULL, "EVP_CIPHER_CTX_new");

  res = EVP_EncryptInit_ex(ctx, cipher, NULL, key, NULL);
  openssl_assert(res == 1, "EVP_EncryptInit_ex");

  res = EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, 12, NULL);
  openssl_assert(res == 1, "EVP_CTRL_AEAD_SET_IVLEN");

  res = EVP_DecryptInit_ex(ctx, NULL, NULL, NULL, nonce);
  openssl_assert(res == 1, "EVP_DecryptInit_ex");

  res = EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, 16, ct + 16);
  openssl_assert(res == 1, "EVP_CTRL_AEAD_SET_TAG");

  res = EVP_CipherUpdate(ctx, NULL, &len, aad, 16);
  openssl_assert(res == 1, "EVP_CipherUpdate(aad)");

  unsigned char *out_buf = malloc(16);
  len = 16;
  res = EVP_CipherUpdate(ctx, out_buf, &len, ct, 16);
  openssl_assert(res == 1, "EVP_CipherUpdate(ct)");
  openssl_assert(len == 16, "EVP_CipherUpdate(ct)");

  unsigned char final_buf[32];
  len = 32;
  res = EVP_CipherFinal(ctx, final_buf, &len);
  openssl_assert(res == 1, "EVP_CipherFinal");
  openssl_assert(len == 0, "EVP_CipherFinal");

  openssl_assert(memcmp(out_buf, pt, 16) == 0, "memcmp");

  printf("Success\n");
}
```

If built on run, on some systems this will pass, on others `EVP_CipherFinal` will fail (I've personally seen that it fails on x86-64 with assembly, and passes on ARM64 and x86-64 without assembly.)

With some ciphers using a ctx in both directions works, because some ciphers only use the raw block cipher in a single encrypt mode, even for decrypt operations (e.g., CTR mode).

To avoid this potential footgun, I recommend one of the following (no opinion offered on which is best):

- Return an error in `EVP_DecryptInit_ex`/`EVP_EncryptInit_ex` anytime the direction is being switched
- Return an error in `EVP_DecryptInit_ex`/`EVP_EncryptInit_ex` where it'd break with the underlying cipher/implementation
- Return the correct result in all cases

**Comment by paulidale** [comment]
I am rather surprised that mixing operations like you are doing comes close to remotely working ever.
This is definite the caller abusing the APIs and OpenSSL has historically not checked for such.
Times are changing and we have been introducing NULL checks into public APIs e.g.

There is a ticket somewhere to enforce algorithm life cycles which ought to catch this too.  I can't seem to locate a reference to it unfortunately.

**Comment by t8m** [comment]
IMO it should be possible to switch an EVP_CIPHER_CTX from encrypt to decrypt operation and back but only with the EVP_EncryptInit_ex/EVP_DecryptInit_ex calls where the key parameter is non-NULL. Otherwise the result is undefined.

Perhaps we should document this for start.

**Comment by t8m** [comment]
I'm just interested to know - if you move setting the key from the first init to the second, does it work for you consistently & correctly?

**Comment by alex** [comment]
Yes, if the key is passed in the second call to `Init` then this passes.

Note that this defeats the point of reusing the `ctx`, which is to reuse the AES key schedule and thus improve performance.

**Comment by t8m** [comment]
Except there is a different key schedule for encryption and decryption for AES. So it cannot really work like this.

**Comment by alex** [comment]
Except, as noted, for ciphers that only use the underlying block cipher in a single direction.

**Comment by t8m** [comment]
Yes, but then this logic cannot be really universal which means it needs to be handled inside the provider implementation and that has its consequences around the complexity of the necessary changes.

**Comment by alex** [comment]
Sure. I don't have a strong perspective, just that if it's not expected to work, it should be an error and not silent misbehavior.

**Comment by t8m** [comment]
> Except, as noted, for ciphers that only use the underlying block cipher in a single direction.

In particular, OCB mode is not one of these so this is probably the main reason why this does not work.

This will not be an easy fix, especially if currently this calling sequence works for the cipher modes that use only the single direction. Because a trivial fix that would disable this would break existing working use-cases of this calling sequence.
